common/s3util.cpp

```diff
@@ -488,9 +489,160 @@ shared_ptr<S3Util> S3Util::BuildS3Util(
+        << "Failed to download file. s3 bucket: "  << th->GetBucketName()
+        << ", s3 key: " << th->GetKey()
+        << ", target_path: " << th->GetTargetFilePath()
+        << ", error: " << th->GetLastError().GetMessage();
+    };
+  context_ = Aws::MakeShared<Aws::Client::AsyncCallerContext>("upload");
+  transfer_manager_ = Aws::Transfer::TransferManager::Create(transfer_config_);
+}
+
+S3Concurrent::~S3Concurrent() {
+    S3Util::TryAwsShutdownAPI();
+}
+
+bool S3Concurrent::setUploadMBps(uint64_t upload_MBps) {
+  // rate limiter is currently setup only with creation of
+  // S3Concurrent object.  rate limit changes via this function only
+  // serve to adjust rate if an initial non-zero rate was set (the
+  // default case).
+  const uint64_t MB = 1000*1000;
+  if (rate_limiter_){
+    uint64_t new_rate_MB = (upload_MBps > 0)? upload_MBps: default_s3_upload_MBps;
+    rate_limiter_->SetRate(new_rate_MB * MB);
+    return true;
+  }
+  return false;
+}
+
+bool S3Concurrent::enqueuePutObject(const string& s3_bucket, const string& key,
+                                    const string& local_path, const string& sync_group_id){
+  LOG(INFO) << "S3Concurrent::enqueuePutObject" << " key: " << key;
+  {
+    std::lock_guard<std::mutex> lock(handles_mutex_);
+    auto sync_collection_ptr = handles_.find(sync_group_id);
+    if (sync_collection_ptr == handles_.end()) {
+      handles_[sync_group_id] = vector<std::shared_ptr<Aws::Transfer::TransferHandle>>();
+    }
+    handles_[sync_group_id].push_back(
+        std::move(transfer_manager_->UploadFile(local_path,
+                                                s3_bucket,
+                                                key,
+                                                "application/octet-stream",
+                                                Aws::Map<Aws::String, Aws::String>(),
+                                                context_)));
+  }
+  return true;
+}
+
+bool S3Concurrent::Sync(const string& sync_group_id) {
+  uint64_t num_files_transferred = 0;
+  uint64_t num_bytes_transferred = 0;
+  uint64_t num_failures = 0;
+  LOG(INFO) << "S3Concurrent::Sync()";
+  vector<std::shared_ptr<Aws::Transfer::TransferHandle>> sync_group;
+  {
+    std::lock_guard<std::mutex> lock(handles_mutex_);
+    auto sync_group_ptr = handles_.find(sync_group_id);
+    if (sync_group_ptr == handles_.end()) {
+      return true;
+    }
+    sync_group.swap(sync_group_ptr->second);
+    handles_.erase(sync_group_ptr);
+  }
+  
+  for(auto &h: sync_group){
+    h->WaitUntilFinished();
+    if (h->GetStatus() == Aws::Transfer::TransferStatus::COMPLETED) {
+      LOG(INFO) << "S3Concurrent::Sync success: " << h->GetTargetFilePath();
+      Stats::get()->Incr(kS3PutObject);
+      num_bytes_transferred += h->GetBytesTransferred();
+      num_files_transferred ++;
+    } else {
+      LOG(INFO) << "S3Concurrent::Sync fail "<< h->GetTargetFilePath();
+      failures_.push_back(h);
+      num_failures++;
```

Comment:
Can u lint the code changes from your editor? 
e.g. the `num_files_transferered ++`, but, `num_failures++`, the former has an extra space before `++`